docs/modules/prepro.rst

```diff
@@ -103,10 +105,166 @@ Some of the code in this package are borrowed from Keras.
    sequences_get_mask
 
 
+..
+  Threading
+  ------------
+  .. autofunction:: threading_data
+
+
+Affine Transform
+----------------
+
+
+Python can be FAST
+^^^^^^^^^^^^^^^^^^
+
+Image augmentation is a critical step in deep learning.
+Though TensorFlow has provided ``tf.image``,
+image augmentation often remains as a key bottleneck.
+``tf.image`` has three limitations:
+
+- Real-world visual tasks such as object detection, segmentation, and pose estimation
+must cope with image meta-data (e.g., coordinates).
+These data are beyond ``tf.image``
+which processes images as tensors.
+
+- ``tf.image`` operators
+breaks the pure Python programing experience (i.e., users have to
+use ``tf.py_func`` in order to call image functions written in Python); however,
+frequent uses of ``tf.py_func`` slow down TensorFlow,
+making users hard to balance flexibility and performance.
+
+- ``tf.image`` API is inflexible. Image operations are
+performed in an order. They are hard to jointly optimize. More importantly,
```

Comment:
I think the major reason for combined affine transformation being fast is because it really has lower computational complexity.

Assume we have $k$ affine transformations $T_i$ in $\mathbb R^2$, where $T_i$ can be represented by 3x3 matrixes.

The sequential transformation can be represented as 

```latex
y = T_k (T_{k-1} \dots T_1 (x))
```

and the time complexity is `O(k N)` where `N` is the cost of apply one transformation to image `x`.
N is linear to the size of x.

For the combined transformation

```latex
y = (T_k \circ \cdots \T_1) (x) 
```

the time complexity is `O(27(k - 1) + N) = O(N)` where 27 = 3^3 is the cost for combine two transformations.

Reply:
Could you add this text into the prepro.rst? I like the way you analyse the time complexity.

Reply:
@lgarithm @zsdonghao